exchanges/binance/binance_websocket.go

```diff
@@ -454,66 +455,158 @@ func (b *Binance) UpdateLocalBuffer(wsdp *WebsocketDepthStream) error {
 		return err
 	}
 
-	currencyPair, err := currency.NewPairFromFormattedPairs(wsdp.Pair,
+	currencyPair, err := currency.NewPairFromFormattedPairs(u.Pair,
 		enabledPairs,
 		format)
 	if err != nil {
 		return err
 	}
 
-	currentBook := b.Websocket.Orderbook.GetOrderbook(currencyPair, asset.Spot)
+	return b.CheckAndLoadBook(currencyPair, asset.Spot, u)
+}
+
+// CheckAndLoadBook buffers updates, spawns workers for http book request and
+// lets slip routine, validates buffer contents and returns required work when
+// needed
+func (b *Binance) CheckAndLoadBook(c currency.Pair, ass asset.Item, u *WebsocketDepthStream) error {
+	b.mtx.Lock() // protects fetching book
+	defer b.mtx.Unlock()
+
+	ch, ok := b.pipe[c.String()]
+	if !ok {
+		ch = make(chan *WebsocketDepthStream, 1000)
+		b.pipe[c.String()] = ch
+	}
+	select {
+	case ch <- u:
+	default:
+		return fmt.Errorf("channel blockage %s", c)
+	}
+
+	if b.fetchingbook[c.String()] {
+		return nil
+	}
+
+	return b.BleedPipe(c, ass)
+}
+
+// BleedPipe applies the buffer to the orderbook
+func (b *Binance) BleedPipe(cp currency.Pair, a asset.Item) error {
+	currentBook := b.Websocket.Orderbook.GetOrderbook(cp, a)
 	if currentBook == nil {
-		// Used when a pair/s is enabled while connected
-		err = b.SeedLocalCache(currencyPair)
-		if err != nil {
-			return err
+		b.initialSync[cp.String()] = true
+		b.fetchingbook[cp.String()] = true
+		select {
+		case b.syncMe <- cp:
+		default:
+			return errors.New("book synchronisation channel blocked up")
 		}
-		currentBook = b.Websocket.Orderbook.GetOrderbook(currencyPair, asset.Spot)
+		return nil
 	}
 
-	// Drop any event where u is <= lastUpdateId in the snapshot.
-	// The first processed event should have U <= lastUpdateId+1 AND u >= lastUpdateId+1.
-	// While listening to the stream, each new event's U should be equal to the previous event's u+1.
-	if wsdp.LastUpdateID <= currentBook.LastUpdateID {
-		return nil
+loop:
+	for {
+		select {
+		case d := <-b.pipe[cp.String()]:
+			if d.LastUpdateID <= currentBook.LastUpdateID {
+				// Drop any event where u is <= lastUpdateId in the snapshot.
+				continue
+			}
+			id := currentBook.LastUpdateID + 1
+			if b.initialSync[cp.String()] {
+				// The first processed event should have U <= lastUpdateId+1 AND
+				// u >= lastUpdateId+1.
+				if d.FirstUpdateID > id && d.LastUpdateID < id {
+					return errors.New("initial sync failure")
+				}
+				b.initialSync[cp.String()] = false
+			} else {
+				// While listening to the stream, each new event's U should be
+				// equal to the previous event's u+1.
+				if d.FirstUpdateID != id {
+					return errors.New("synchronisation failure")
```

Comment:
This definitely did no enjoy being disconnected and reconnected. It makes no attempt at resubscribing and will constantly pass this error to the datahandler.

It also couldn't handle disabling a pair and re-enabling. I think its not clearing the orderbook when that occurs